flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -669,13 +677,18 @@ private void stopTaskExecutorServices() throws Exception {
                             taskInformation.getJobVertexId(),
                             tdd.getSubtaskIndex());
 
+            final StateChangelogStorage<?> changelogStorage =
+                    changelogStoragesManager.stateChangelogStorageForJob(
+                            jobId, jobInformation.getJobConfiguration());
```

Comment:
Just a question: do we need to merge job configuration with task configuration here?

Reply:
I checked the code and IIUC, the configuration in ```taskInformation.getConfiguration()``` comes from the ```JobVertex.getConfiguration()``` and be set in ```TaskConfig```. It seems that no checkpoint related configs should be set here. 

As for the ```TaskExecutor.taskManagerConfiguration```, it is a configuration for the cluster. Since there is a default value for changelog storage, the job configuration will always override the one for cluster. I don't think we need any configuration overriding here.

WDYT?

Reply:
I think we are currently missing the API to set `STATE_CHANGE_LOG_STORAGE` from the user program (similar to `enableChangelogStateBackend()` for `ENABLE_STATE_CHANGE_LOG`).

If we add it (and implement similarly) then the value will come in `taskInformation.getTaskConfiguration()`.

And the value configured in yaml will come in `taskManagerConfiguration.getConfiguration()` as you pointed out.
So `taskInformation.getTaskConfiguration()` should override `taskManagerConfiguration.getConfiguration()` - if the API is added.

From the consistency point of view, the API should be added in this PR (something like `enableChangelogStateBackend("memory")`).

Reply:
I agree that we could add a datastream API for ```STATE_CHANGE_LOG_STORAGE```. We could do this later when we have several implementation of storage.
However, I prefer make another API instead of ```enableChangelogStateBackend("memory")```. It's not user-friendly to make user know the storage shotcut name before using changelog, especially we may have only one storage for production in most cases.

Reply:
Sure, I meant a separate (overloaded) method, so users would have both. Sorry for the confusion.

Reply:
Yeah, this seems good.

Reply:
According to offline sync, we:
- currently reads value from taskmanager configuration
- pass value from user program and do overriding here